### PR TITLE
Querytee: Add metric to measure relative backend latency

### PR DESCRIPTION
**Pull request by jhesketh**
This metric gives us a measurement of how individual queries compare between two backends in terms of latency (or duration).

#### Which issue(s) this PR fixes or relates to

Fixes #<issue number>

#### Checklist

- [x] Tests updated.
- [x] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: tools/querytee/proxy_metrics.go [hunk 1]
```diff
@@ -43,6 +44,12 @@ func NewProxyMetrics(registerer prometheus.Registerer) *ProxyMetrics {
 			Name:      "responses_compared_total",
 			Help:      "Total number of responses compared per route name by result.",
 		}, []string{"route", "result"}),
+		relativeDuration: promauto.With(registerer).NewHistogramVec(prometheus.HistogramOpts{
+			Namespace: queryTeeMetricsNamespace,
+			Name:      "backend_request_relative_duration_seconds",
+			Help:      "Time (in seconds) of preferred backend less secondary backend.",
+			Buckets:   []float64{-100, -50, -25, -10, -5, -4, -3, -2, -1.5, -1, -.75, -.5, -.25, -.1, -.05, -.025, -.01, -.005, .005, .01, .025, .05, .1, .25, .5, 0.75, 1, 1.5, 2, 3, 4, 5, 10, 25, 50, 100},
+		}, []string{"method", "route"}),
```
Comment (charleskorn) [inline tools/querytee/proxy_metrics.go:52]: This might be a good candidate for a native histogram.

File: tools/querytee/proxy_endpoint.go [hunk 2]
```diff
@@ -224,6 +224,8 @@ func (p *ProxyEndpoint) executeBackendRequests(req *http.Request, resCh chan *ba
 			)
 		}
 
+		relativeDuration := expectedResponse.elapsedTime - actualResponse.elapsedTime
+		p.metrics.relativeDuration.WithLabelValues(req.Method, p.routeName).Observe(relativeDuration.Seconds())
```
Comment (charleskorn) [inline tools/querytee/proxy_endpoint.go:228]: Do we need to differentiate between HTTP methods here? Performance should be the same regardless of whether it's a GET or a POST request.

Comment (charleskorn) [inline tools/querytee/proxy_endpoint.go:228]: It might be worth including the comparison result as a label on this metric - I could imagine these results being skewed if one backend is quickly returning a lot of errors, for example.

**Review by charleskorn — COMMENTED** [review]
Thanks for working on this.

It'd be good to add the proportional latency difference as well - a request being 1s slower is interesting, but knowing whether that's a 1% change or a 100% change would be useful as well.

**Review by pracucci — COMMENTED** [review]
> This metric gives us a measurement of how individual queries compare between two backends in terms of latency (or duration).

Why do we need a new metric? Can't we just solve this "problem" at PromQL level, with a query?

File: tools/querytee/proxy_endpoint.go (hunk 2 above)
Comment (jhesketh) [reply tools/querytee/proxy_endpoint.go:228]: Good point.

Comment (jhesketh) [reply tools/querytee/proxy_endpoint.go:228]: Great idea, will do

**Comment by jhesketh** [comment]
> Why do we need a new metric? Can't we just solve this "problem" at PromQL level, with a query?

We currently measure the `backend_request_duration_seconds` as a histogram with a label for the backend. We can subtract one backend from the other to get a general indication of which one is more performant, but it isn't necessarily accurate as results can cancel each other out.

By tracking with a new metric, we can see on this individual histogram the number of times one backend is faster than the other (and by how much) depending if the bucket is positive or negative.

File: tools/querytee/proxy_metrics.go (hunk 1 above)
Comment (jhesketh) [reply tools/querytee/proxy_metrics.go:52]: I don't think negatives can be used in native histograms (correct me if I'm wrong). In which case we'd need two metrics: one for when the primary is more performant, and one for the secondary. WDYT?

Comment (jhesketh) [reply tools/querytee/proxy_metrics.go:52]: Ah, then I am mistaken. I will try this then, thanks.

File: tools/querytee/proxy_endpoint_test.go [hunk 3]
```diff
@@ -353,67 +354,81 @@ func Test_ProxyEndpoint_Comparison(t *testing.T) {
 func Test_ProxyEndpoint_LogSlowQueries(t *testing.T) {
 	scenarios := map[string]struct {
 		slowResponseThreshold         time.Duration
-		preferredResponseLatency      time.Duration
-		secondaryResponseLatency      time.Duration
+		latencyPairs                  []latencyPair
```
Comment (charleskorn) [inline tools/querytee/proxy_endpoint_test.go:357]: Is this change necessary? The previous setup seems clearer to me, and we can construct the `[]time.Duration` slices for `newMockProxyBackend` directly (rather than using `splitLatencyPairs`)

File: tools/querytee/proxy_metrics.go [hunk 4]
```diff
@@ -45,11 +46,17 @@ func NewProxyMetrics(registerer prometheus.Registerer) *ProxyMetrics {
...
-		}, []string{"method", "route"}),
+			Namespace:                   queryTeeMetricsNamespace,
+			Name:                        "backend_request_relative_duration_seconds",
+			Help:                        "Time (in seconds) of preferred backend less secondary backend.",
+			NativeHistogramBucketFactor: 2,
+		}, []string{"route"}),
+		proportionalDuration: promauto.With(registerer).NewHistogramVec(prometheus.HistogramOpts{
+			Namespace:                   queryTeeMetricsNamespace,
+			Name:                        "backend_request_relative_duration_proportional",
+			Help:                        "Proportional time (%) of preferred backend vs secondary backend.",
```
Comment (charleskorn) [inline tools/querytee/proxy_metrics.go:57]: [nit] Perhaps this would be clearer?

```suggestion
			Help:                        "Response time of preferred backend, as a proportion of secondary backend response time.",
```

File: tools/querytee/proxy_endpoint_test.go [hunk 5]
```diff
@@ -793,11 +780,20 @@ func (b *mockProxyBackend) Preferred() bool {
 }
 
 func (b *mockProxyBackend) ForwardRequest(_ *http.Request, _ io.ReadCloser) (time.Duration, int, []byte, *http.Response, error) {
+	statusCode := 200
+	if b.responseIndex >= len(b.fakeResponseLatencies) {
+		statusCode = 500
+	}
```
Comment (charleskorn) [inline tools/querytee/proxy_endpoint_test.go:786]: [nit] We could set the status code on the response to 200 by default, then change it in the second `if b.responseIndex >= ...` block below - this would help keep the condition in one place

**Review by charleskorn — COMMENTED** [review]
Overall LGTM, modulo comment on test changes.

Could you please add a changelog entry?

File: tools/querytee/proxy_endpoint_test.go (hunk 3 above)
Comment (jhesketh) [reply tools/querytee/proxy_endpoint_test.go:357]: This lets us create multiple data points for the histogram. Possibly overkill, but it's here now, so unless you think it's overly convoluted I'd say we leave it in case we need it for future tests.

File: docs/sources/mimir/manage/tools/query-tee.md [hunk 6]
```diff
@@ -163,6 +163,11 @@ cortex_querytee_responses_total{backend="<hostname>",method="<method>",route="<r
 cortex_querytee_responses_compared_total{route="<route>",result="<success|fail>"}
 ```
 
+Additionally, if backend results comparison is configured, two native-histograms are available:
+
+* `cortex_querytee_backend_request_relative_duration_seconds`: Time (in seconds) of preferred backend less secondary backend.
```
Comment (jdbaldry) [inline docs/sources/mimir/manage/tools/query-tee.md:168]: Maybe rephrase to make it clear that it's the difference between the two?

```suggestion
* `cortex_querytee_backend_request_relative_duration_seconds`: The difference of time, in seconds, between the preferred backend request duration and the secondary backend request duration.
```

File: docs/sources/mimir/manage/tools/query-tee.md [hunk 7]
```diff
@@ -163,6 +163,11 @@ cortex_querytee_responses_total{backend="<hostname>",method="<method>",route="<r
...
+* `cortex_querytee_backend_request_relative_duration_proportional`: Response time of preferred backend, as a proportion of secondary backend response time.
```
Comment (jdbaldry) [inline docs/sources/mimir/manage/tools/query-tee.md:169]: Should it be request rather than response given the metric name?

```suggestion
* `cortex_querytee_backend_request_relative_duration_proportional`: The request duration of the preferred backend as a proportion of secondary backend request duration.
```

File: tools/querytee/proxy_endpoint_test.go (hunk 3 above)
Comment (charleskorn) [reply tools/querytee/proxy_endpoint_test.go:357]: For `Test_ProxyEndpoint_RelativeDurationMetric`, I think using `latencyPairs` like this makes sense, but for this test (`Test_ProxyEndpoint_LogSlowQueries`), it obscures the intent - we never intend to test multiple requests in this test.

File: docs/sources/mimir/manage/tools/query-tee.md [hunk 8]
```diff
@@ -163,6 +163,11 @@ cortex_querytee_responses_total{backend="<hostname>",method="<method>",route="<r
 cortex_querytee_responses_compared_total{route="<route>",result="<success|fail>"}
 ```
 
+Additionally, if backend results comparison is configured, two native-histograms are available:
```
Comment (charleskorn) [inline docs/sources/mimir/manage/tools/query-tee.md:166]: [nit] No hyphen required

```suggestion
Additionally, if backend results comparison is configured, two native histograms are available:
```

File: tools/querytee/proxy_endpoint_test.go (hunk 3 above)
Comment (jhesketh) [reply tools/querytee/proxy_endpoint_test.go:357]: There is no reason why `Test_ProxyEndpoint_LogSlowQueries` couldn't test multiple points though. It also verifies the number of points with `waitForResponseComparisonMetric` which takes in the expected count.

I can change it back if you think it's a big issue, but in my opinion this makes the mock asset more flexible for future tests etc. and would rather not waste too many cycles here.

File: docs/sources/mimir/manage/tools/query-tee.md (hunk 6 above)
Comment (jhesketh) [reply docs/sources/mimir/manage/tools/query-tee.md:168]: This is copied from the help string of the new metric where the description is consistent with other metrics.

File: docs/sources/mimir/manage/tools/query-tee.md (hunk 7 above)
Comment (jhesketh) [reply docs/sources/mimir/manage/tools/query-tee.md:169]: This should be response time. I've updated the metric names to be clearer.

**Review by jdbaldry — APPROVED** [review]
Thanks for updating the metric name.

I think consistency with other metric descriptions is probably preferable to more clarity.

File: tools/querytee/proxy_endpoint_test.go [hunk 9]
```diff
@@ -446,16 +447,106 @@ func Test_ProxyEndpoint_LogSlowQueries(t *testing.T) {
...
+
+			gotProportional := filterMetrics(got, []string{"cortex_querytee_backend_response_relative_duration_proportional"})
+			require.Equal(t, 1, len(gotProportional), "Expect only one metric after filtering")
+			require.Equal(t, scenario.expectedSampleCount, gotProportional[0].Metric[0].Histogram.GetSampleCount())
+			require.Equal(t, scenario.expectedProportionalSampleSum, gotProportional[0].Metric[0].Histogram.GetSampleSum())
+		})
+	}
+}
+
+func filterMetrics(metrics []*dto.MetricFamily, names []string) []*dto.MetricFamily {
```
Comment (charleskorn) [inline tools/querytee/proxy_endpoint_test.go:527]: [nit] This method is only ever used with one value for `name` - might be able to simplify this and the test that calls it.

**Review by charleskorn — APPROVED** [review]
